File: tests/fixtures/user/linux/src/sysvshm.c
/* System V shared memory (ipc/shm.c). A segment's creation and status;
 * attaches seeing each other's stores; attaches counted by mapping (a split
 * counts twice, and is not merged again); shmdt and munmap detaching; a forked child's inherited
 * attach and its exit; read-only attaches; SHM_RND and SHM_REMAP; keys
 * (sizes, IPC_EXCL, ENOENT); access for another user; IPC_SET, SHM_STAT,
 * IPC_INFO, SHM_INFO; removal while attached (SHM_DEST, the key private,
 * still attachable by identifier) and at the last detach, also a dying
 * process's; /proc/self/maps. Values that depend on other segments in the
 * namespace are not printed. */
#define _GNU_SOURCE
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ipc.h>
#include <sys/mman.h>
#include <sys/shm.h>
#include <sys/wait.h>
#include <unistd.h>
#include "check.h"

#define PAGE 4096

static struct shmid_ds st(int id) {
    struct shmid_ds ds;
    memset(&ds, 0xff, sizeof ds);
    if (shmctl(id, IPC_STAT, &ds) != 0)
        ds.shm_segsz = 0;
    return ds;
}

/* The /proc/self/maps line of address a, if any. */
static int maps_line(void *a, char *out, size_t n) {
    FILE *f = fopen("/proc/self/maps", "r");
    char line[512];
    char want[32];
    snprintf(want, sizeof want, "%08lx-", (unsigned long)a);
    int found = 0;
    while (f && fgets(line, sizeof line, f))
        if (!strncmp(line, want, strlen(want))) {
            snprintf(out, n, "%s", line);
            found = 1;
        }
    if (f)
        fclose(f);
    return found;
}

static void lifecycle(void) {
    int id = shmget(IPC_PRIVATE, 3 * PAGE + 100, 0600);
    struct shmid_ds ds = st(id);
    CHECK("create", id >= 0 && ds.shm_segsz == 3 * PAGE + 100 && ds.shm_nattch == 0 &&
                        ds.shm_cpid == getpid() && ds.shm_lpid == 0 && ds.shm_atime == 0 &&
                        ds.shm_dtime == 0 && ds.shm_ctime > 0 && ds.shm_perm.mode == 0600 &&
                        ds.shm_perm.__key == IPC_PRIVATE && ds.shm_perm.uid == geteuid() &&
                        ds.shm_perm.cuid == geteuid());
    char *a = shmat(id, NULL, 0), *b = shmat(id, NULL, 0);
    CHECK("attach", a != (void *)-1 && b != (void *)-1 && a != b);
    a[3 * PAGE + 99] = 'x';
    CHECK("shared-stores", b[3 * PAGE + 99] == 'x');
    ds = st(id);
    CHECK("attach-count", ds.shm_nattch == 2 && ds.shm_lpid == getpid() && ds.shm_atime > 0);
    mprotect(a, PAGE, PROT_READ);
    CHECK("split-counts-twice", st(id).shm_nattch == 3);
    /* The pieces are not merged again (is_mergeable_vma: shm's VMAs have a
     * close operation). */
    mprotect(a, PAGE, PROT_READ | PROT_WRITE);
    CHECK("no-merge", st(id).shm_nattch == 3);
    char line[512];
    CHECK("maps", maps_line(a, line, sizeof line) && strstr(line, " rw-s 00000000 00:01 ") &&
                      strstr(line, "/SYSV00000000 (deleted)"));
    char ino[32];
    snprintf(ino, sizeof ino, " %d ", id);
    CHECK("maps-inode-is-id", strstr(line, ino) != NULL);
    CHECK("shmdt", shmdt(a) == 0 && st(id).shm_nattch == 1 && st(id).shm_dtime > 0);
    CHECK_ERR("shmdt-again", shmdt(a), EINVAL);
    CHECK_ERR("shmdt-unaligned", shmdt(b + 1), EINVAL);

    /* A forked child's attach is its own; its exit detaches it. */
    pid_t c = fork();
    if (c == 0) {
        int ok = b[3 * PAGE + 99] == 'x' && st(id).shm_nattch == 2;
        b[0] = 'c';
        _exit(ok ? 0 : 1);
    }
    int status = -1;
    waitpid(c, &status, 0);
    ds = st(id);
    CHECK("child-attach", WIFEXITED(status) && WEXITSTATUS(status) == 0);
    CHECK("child-exit-detaches", ds.shm_nattch == 1 && ds.shm_lpid == c && b[0] == 'c');

    CHECK("munmap-detaches", munmap(b, 4 * PAGE) == 0 && st(id).shm_nattch == 0);
    char *r = shmat(id, NULL, SHM_RDONLY);
    CHECK("read-only", r != (void *)-1 && r[3 * PAGE + 99] == 'x');
    CHECK_ERR("read-only-stays", mprotect(r, PAGE, PROT_READ | PROT_WRITE), EACCES);
    shmdt(r);

    /* Placement. */
    char *hole = mmap(NULL, 8 * PAGE, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    munmap(hole, 8 * PAGE);
    CHECK_ERR("unaligned", (long)shmat(id, hole + 1, 0), EINVAL);
    CHECK("rounded", shmat(id, hole + 1, SHM_RND) == hole);
    CHECK_ERR("occupied", (long)shmat(id, hole, 0), EINVAL);
    CHECK("remap", shmat(id, hole, SHM_REMAP) == hole);
    CHECK_ERR("remap-needs-address", (long)shmat(id, NULL, SHM_REMAP), EINVAL);
    shmdt(hole);
    CHECK_ERR("no-such-id", (long)shmat(id + 1, NULL, 0), EINVAL);
    CHECK_ERR("negative-id", (long)shmat(-1, NULL, 0), EINVAL);

    /* Removal while attached. */
    a = shmat(id, NULL, 0);
    CHECK("rmid-attached", shmctl(id, IPC_RMID, NULL) == 0);
    ds = st(id);
    CHECK("rmid-marks", ds.shm_perm.mode == (0600 | 01000) && ds.shm_perm.__key == IPC_PRIVATE);
    b = shmat(id, NULL, 0);
    CHECK("attach-marked", b != (void *)-1 && b[0] == 'c');
    shmdt(a);
    CHECK("still-there", st(id).shm_nattch == 1);
    shmdt(b);
    CHECK_ERR("gone-at-last-detach", shmctl(id, IPC_STAT, &ds), EINVAL);
}

static void keys(void) {
    key_t key = 0x52410000 ^ getpid();
    int id = shmget(key, 100, IPC_CREAT | 0640);
    CHECK("key-create", id >= 0);
    CHECK("key-find", shmget(key, 50, 0) == id);
    CHECK_ERR("key-too-big", shmget(key, 101, 0), EINVAL);
    CHECK_ERR("key-exclusive", shmget(key, 1, IPC_CREAT | IPC_EXCL | 0600), EEXIST);
    CHECK_ERR("key-missing", shmget(key ^ 1, 1, 0), ENOENT);
    CHECK_ERR("size-zero", shmget(IPC_PRIVATE, 0, 0600), EINVAL);
    struct shmid_ds ds;
    CHECK("shm-stat-by-index", shmctl(id & 0x7fff, SHM_STAT, &ds) == id && ds.shm_perm.__key == key);
    ds = st(id);
    ds.shm_perm.mode = 0604;
    CHECK("ipc-set", shmctl(id, IPC_SET, &ds) == 0 && st(id).shm_perm.mode == 0604);
    struct shminfo info;
    CHECK("ipc-info", shmctl(0, IPC_INFO, (struct shmid_ds *)&info) >= 0 && info.shmmni == 4096 &&
                          info.shmmin == 1);
    struct shm_info si;
    CHECK("shm-info", shmctl(0, SHM_INFO, (struct shmid_ds *)&si) >= 0 && si.used_ids >= 1);
    CHECK_ERR("bad-command", shmctl(id, 99, &ds), EINVAL);
    CHECK_ERR("bad-buffer", shmctl(id, IPC_STAT, (struct shmid_ds *)8), EFAULT);
    shmctl(id, IPC_RMID, NULL);

    /* Another (unprivileged) user and a segment no one may use. */
    int locked = shmget(IPC_PRIVATE, 1, 0);
    pid_t c = fork();
    if (c == 0) {
        /* Root becomes another user; another user stays the owner, who
         * has no access to a mode-0 segment either but may remove it. */
        int other = geteuid() == 0;
        if (other && (setgid(65534) || setuid(65534)))
            _exit(2);
        int r = shmat(locked, NULL, SHM_RDONLY) == (void *)-1 && errno == EACCES &&
                (!other || (shmctl(locked, IPC_RMID, NULL) == -1 && errno == EPERM));
        _exit(r ? 0 : 1);
    }
    int status = -1;
    waitpid(c, &status, 0);
    CHECK("other-user", WIFEXITED(status) && WEXITSTATUS(status) == 0);
    shmctl(locked, IPC_RMID, NULL);

    /* A process that dies attached to a removed segment removes it. */
    int dying = shmget(IPC_PRIVATE, PAGE, 0600);
    c = fork();
    if (c == 0) {
        shmat(dying, NULL, 0);
        shmctl(dying, IPC_RMID, NULL);
        _exit(0);
    }
    waitpid(c, &status, 0);
    struct shmid_ds gone;
    CHECK_ERR("removed-at-exit", shmctl(dying, IPC_STAT, &gone), EINVAL);
}

int main(void) {
    setvbuf(stdout, NULL, _IOLBF, 0);
    lifecycle();
    keys();
    FINISH();
}

File: src/user/linux/tests/sysvshm.rs
//! System V shared memory against `ipc/shm.c` (Linux 6.19), through the
//! system calls on every ABI: `shmget`, `shmat` (placement, `SHM_RND`,
//! `SHM_REMAP`, `SHM_RDONLY`), `shmdt`, `shmctl`, attach counting by
//! mapping, removal while attached, and `/proc/self/maps`. Each harness
//! has a namespace of its own.

use super::harness::{Harness, P, each_abi};
use crate::user::linux::abi::Sysno;
use crate::user::linux::abi::errno_table::*;

const IPC_PRIVATE: u64 = 0;
const IPC_CREAT: u64 = 0o1000;
const IPC_EXCL: u64 = 0o2000;
const IPC_RMID: u64 = 0;
const IPC_SET: u64 = 1;
const IPC_STAT: u64 = 2;
const IPC_INFO: u64 = 3;
const SHM_STAT: u64 = 13;
const SHM_INFO: u64 = 14;
const SHM_LOCK: u64 = 11;
const SHM_RDONLY: u64 = 0o10000;
const SHM_RND: u64 = 0o20000;
const SHM_REMAP: u64 = 0o40000;
const SHM_DEST: u32 = 0o1000;
const SHM_LOCKED: u32 = 0o2000;
const PROT_READ: u64 = 1;
const PROT_WRITE: u64 = 2;

fn get(h: &Harness, at: u64, n: usize) -> Vec<u8> {
    let mut b = vec![0u8; n];
    h.proc.state.space.read(at, &mut b).unwrap();
    b
}

fn u64_at(b: &[u8], at: usize) -> u64 {
    u64::from_le_bytes(b[at..at + 8].try_into().unwrap())
}

fn u32_at(b: &[u8], at: usize) -> u32 {
    u32::from_le_bytes(b[at..at + 4].try_into().unwrap())
}

/// `struct shmid64_ds` of `id`.
fn stat(h: &mut Harness, id: u64, buf: u64) -> Vec<u8> {
    h.ok(Sysno::Shmctl, &[id, IPC_STAT, buf]);
    get(h, buf, 112)
}

fn root() -> bool {
    // SAFETY: geteuid has no failure mode.
    unsafe { libc::geteuid() == 0 }
}

#[test]
fn segments_attach_and_detach() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let buf = h.anon(4 * P, 3, false);
        let id = h.ok(Sysno::Shmget, &[IPC_PRIVATE, 5000, 0o600]);
        let ds = stat(&mut h, id, buf);
        assert_eq!(u64_at(&ds, 48), 5000, "shm_segsz");
        assert_eq!(u64_at(&ds, 88), 0, "shm_nattch");
        assert!(u64_at(&ds, 72) > 0, "shm_ctime");
        // Two attaches see each other's stores.
        let a = h.ok(Sysno::Shmat, &[id, 0, 0]);
        let b = h.ok(Sysno::Shmat, &[id, 0, 0]);
        assert_ne!(a, b);
        h.proc.state.space.write(a + 4999, &[0x5A]).unwrap();
        assert_eq!(get(&h, b + 4999, 1), [0x5A]);
        let ds = stat(&mut h, id, buf);
        assert_eq!(u64_at(&ds, 88), 2);
        assert_eq!(u32_at(&ds, 84) as i32, h.proc.state.pid, "shm_lpid");
        assert!(u64_at(&ds, 56) > 0, "shm_atime");
        // A mapping split in two counts twice (shm_open on the split).
        h.ok(Sysno::Mprotect, &[a, P, PROT_READ]);
        assert_eq!(u64_at(&stat(&mut h, id, buf), 88), 3);
        // ... and is not merged again (is_mergeable_vma).
        h.ok(Sysno::Mprotect, &[a, P, PROT_READ | PROT_WRITE]);
        assert_eq!(u64_at(&stat(&mut h, id, buf), 88), 3);
        // shmdt: the attach at that address (all its pieces).
        h.ok(Sysno::Shmdt, &[a]);
        let ds = stat(&mut h, id, buf);
        assert_eq!(u64_at(&ds, 88), 1);
        assert!(u64_at(&ds, 64) > 0, "shm_dtime");
        assert_eq!(h.err(Sysno::Shmdt, &[a]), EINVAL);
        assert_eq!(h.err(Sysno::Shmdt, &[b + 1]), EINVAL);
        // munmap detaches too.
        h.ok(Sysno::Munmap, &[b, 2 * P]);
        assert_eq!(u64_at(&stat(&mut h, id, buf), 88), 0);
        // /proc/self/maps: shmem, the identifier as inode.
        let c = h.ok(Sysno::Shmat, &[id, 0, SHM_RDONLY]);
        let maps = String::from_utf8(crate::user::linux::procfs::maps(&h.proc.state)).unwrap();
        let line = maps
            .lines()
            .find(|l| l.starts_with(&format!("{c:08x}-")))
            .unwrap();
        assert!(line.contains(" r--s 00000000 00:01 "), "{line}");
        assert!(line.ends_with(" /SYSV00000000 (deleted)"), "{line}");
        assert!(line.contains(&format!(" {id} ")), "{line}");
        // Read-only: never writable.
        assert_eq!(h.err(Sysno::Mprotect, &[c, P, PROT_WRITE]), EACCES);
        h.ok(Sysno::Shmdt, &[c]);
    });
}

#[test]
fn shmat_places_as_do_shmat() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let id = h.ok(Sysno::Shmget, &[IPC_PRIVATE, 2 * P, 0o600]);
        let free = h.anon(4 * P, 3, false);
        h.ok(Sysno::Munmap, &[free, 4 * P]);
        // An unaligned address: EINVAL, or rounded down with SHM_RND.
        assert_eq!(h.err(Sysno::Shmat, &[id, free + 1, 0]), EINVAL);
        assert_eq!(h.call(Sysno::Shmat, &[id, free + 1, SHM_RND]), free as i64);
        // Over a mapping: EINVAL, unless SHM_REMAP.
        assert_eq!(h.err(Sysno::Shmat, &[id, free, 0]), EINVAL);
        assert_eq!(h.call(Sysno::Shmat, &[id, free, SHM_REMAP]), free as i64);
        assert_eq!(h.err(Sysno::Shmat, &[id, 0, SHM_REMAP]), EINVAL);
        assert_eq!(h.err(Sysno::Shmat, &[id, 1, SHM_RND | SHM_REMAP]), EINVAL);
        // Unknown and negative identifiers.
        assert_eq!(h.err(Sysno::Shmat, &[id + 1, 0, 0]), EINVAL);
        assert_eq!(h.err(Sysno::Shmat, &[u32::MAX as u64, 0, 0]), EINVAL);
    });
}

#[test]
fn shmctl_follows_ksys_shmctl() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let buf = h.anon(4 * P, 3, false);
        let id = h.ok(Sysno::Shmget, &[42, 100, IPC_CREAT | 0o640]);
        assert_eq!(h.call(Sysno::Shmget, &[42, 50, 0]), id as i64);
        assert_eq!(h.err(Sysno::Shmget, &[42, 101, 0]), EINVAL);
        assert_eq!(h.err(Sysno::Shmget, &[42, 1, IPC_CREAT | IPC_EXCL]), EEXIST);
        assert_eq!(h.err(Sysno::Shmget, &[43, 1, 0]), ENOENT);
        assert_eq!(h.err(Sysno::Shmget, &[IPC_PRIVATE, 0, 0]), EINVAL);
        // SHM_STAT takes an index, and returns the identifier.
        assert_eq!(h.call(Sysno::Shmctl, &[0, SHM_STAT, buf]), id as i64);
        assert!(h.call(Sysno::Shmctl, &[0, IPC_INFO, buf]) >= 0);
        assert_eq!(u64_at(&get(&h, buf, 72), 16), 4096, "shmmni");
        assert!(h.call(Sysno::Shmctl, &[0, SHM_INFO, buf]) >= 0);
        assert_eq!(u32_at(&get(&h, buf, 4), 0), 1, "used_ids");
        assert_eq!(h.err(Sysno::Shmctl, &[id, 99, buf]), EINVAL);
        assert_eq!(h.err(Sysno::Shmctl, &[id, IPC_STAT, 8]), EFAULT);
        assert_eq!(
            h.err(Sysno::Shmctl, &[u32::MAX as u64, IPC_STAT, buf]),
            EINVAL
        );
        // IPC_SET: the permission bits and owner; ctime moves.
        let mut ds = stat(&mut h, id, buf);
        ds[20..24].copy_from_slice(&0o600u32.to_le_bytes());
        h.proc.state.space.write(buf, &ds).unwrap();
        h.ok(Sysno::Shmctl, &[id, IPC_SET, buf]);
        assert_eq!(u32_at(&stat(&mut h, id, buf), 20), 0o600);
        if !root() {
            // No memory-lock limit: SHM_LOCK is refused.
            h.proc.state.rlimits[8] = (0, 0);
            assert_eq!(h.err(Sysno::Shmctl, &[id, SHM_LOCK, 0]), EPERM);
        }
        h.proc.state.rlimits[8] = (65536, 65536);
        h.ok(Sysno::Shmctl, &[id, SHM_LOCK, 0]);
        assert_eq!(u32_at(&stat(&mut h, id, buf), 20) & SHM_LOCKED, SHM_LOCKED);
        // Removal while attached: marked, the key private; the last
        // detach removes it.
        let a = h.ok(Sysno::Shmat, &[id, 0, 0]);
        h.ok(Sysno::Shmctl, &[id, IPC_RMID, 0]);
        let ds = stat(&mut h, id, buf);
        assert_eq!(u32_at(&ds, 20) & SHM_DEST, SHM_DEST);
        assert_eq!(u32_at(&ds, 0), 0, "the key is IPC_PRIVATE");
        assert_eq!(h.err(Sysno::Shmget, &[42, 1, 0]), ENOENT);
        // Still attachable by its identifier.
        let b = h.ok(Sysno::Shmat, &[id, 0, 0]);
        h.ok(Sysno::Shmdt, &[a]);
        h.ok(Sysno::Shmdt, &[b]);
        assert_eq!(h.err(Sysno::Shmctl, &[id, IPC_STAT, buf]), EINVAL);
        // A new segment by the same key is another one.
        let again = h.ok(Sysno::Shmget, &[42, 1, IPC_CREAT | 0o600]);
        assert_ne!(again, id);
    });
}
